crates/bevy_utils/src/staging.rs

```diff
@@ -0,0 +1,1108 @@
+    /// When this is done, there should be no more staged changes, and [`any_staged`](Self::any_staged) should be `false`.
+    fn apply_staged(&mut self, storage: &mut Self::Cold);
+
+    /// Returns `true` if and only if there are staged changes that could be applied.
+    fn any_staged(&self) -> bool;
+}
+
+/// This trait defines relevant types for [`StagableWrites`].
+/// See [`this github issue`](https://github.com/rust-lang/rust/issues/87479) for why this needs to be separate.
+pub trait StagableWritesTypes: Sized {
+    /// This is the type that will store staged changes.
+    type Staging: StagedChanges;
+    /// This is the type that will reference [`Cold`](StagedChanges::Cold) for [`Staging`](Self::Staging).
+    /// This is left generalized so that it can be put in a lock or otherwise if necessary.
+    type ColdRef<'a>: Deref<Target = <Self::Staging as StagedChanges>::Cold>
+    where
+        <Self::Staging as StagedChanges>::Cold: 'a;
+    /// This is the type that will mutably reference [`Cold`](StagedChanges::Cold) for [`Staging`](Self::Staging).
+    /// This is left generalized so that it can be put in a lock or otherwise if necessary.
+    type ColdMut<'a>: Deref<Target = <Self::Staging as StagedChanges>::Cold>
+    where
+        <Self::Staging as StagedChanges>::Cold: 'a;
+}
+
+/// This trait generallizes the stage on write concept.
+pub trait StagableWritesCore: StagableWritesTypes {
+    /// Allows raw access to reading cold storage, which may still have unapplied staged changes that make this out of date.
+    /// Use this to return data attached to a lock guard when one such guard is already in existence.
+    ///
+    /// This must never deadlock if there is already a read for this value on this thread.
+    fn raw_read_cold(&self) -> Self::ColdRef<'_>;
+
+    /// Allows raw access to reading staged changes, which may be missing context of cold storage.
+    /// Use this to return data attached to a lock guard when one such guard is already in existence.
+    ///
+    /// This must never deadlock if there is already a read for this value on this thread.
+    fn raw_read_staged(&self) -> RwLockReadGuard<'_, Self::Staging>;
+
+    /// Same as [`raw_read_cold`](StagableWritesCore::raw_read_cold), but never blocks.
+    fn raw_read_cold_non_blocking(&self) -> Option<Self::ColdRef<'_>>;
+
+    /// Same as [`raw_read_staged`](StagableWritesCore::raw_read_staged), but never blocks.
+    fn raw_read_staged_non_blocking(&self) -> Option<RwLockReadGuard<'_, Self::Staging>>;
+
+    /// Allows raw access to reading staged changes, which may be missing context of cold storage.
+    fn raw_write_staged(&self) -> RwLockWriteGuard<'_, Self::Staging>;
+
+    /// Same as [`raw_write_staged`](StagableWritesCore::raw_write_staged), but never blocks.
+    fn raw_write_staged_non_blocking(&self) -> Option<RwLockWriteGuard<'_, Self::Staging>>;
+
+    /// Allows raw access to both staged and cold data.
```

Comment:
`raw_write_cold_mut` and `raw_write_staged_mut` both mention that they don't lock. Should we add this to the comment here? e.g., `This should not do any locking, and shouldn't need to as we have exclusive references.`